File: experimental/cpp/usb_packet_testing/src/packet_engine.cpp
#include "packet_engine.h"

#include "usb.h"
#include "packet.h"
#include "common.h"
#include "cobs.h"
#include "file_receiver.h"

#include <string.h>

static char s_rx_buf[PACKET_ENGINE_BUFFER_SIZE];
static int s_rx_buf_len = 0;

int handle_received_packet(uint8_t *buf, int len)
{
    // COBS decode
    cobs_decode_result res = cobs_decode(buf, len, buf, len);
    len = res.out_len;

    int err = packet_check(buf, len);
    if (err < 0)
    {
        // Invalid packet
        TRACE_PRINTF("Received invalid packet, error: %d\n", err);

        return -1;
    }

    PacketHeader *header = (PacketHeader *)buf;

    TRACE_PRINTF("Received valid packet type: %d, len: %d\n", header->id.type, header->length);
    TRACE_PRINTF("Packet data:");
    for (int i = 0; i < header->length; i++)
    {
        printf("%c", buf[sizeof(PacketHeader) + i]);
    }
    printf("\n");

    // Send off the packet
    switch (header->id.type)
    {
    case PACKET_TYPE_FILE_START.type:
    case PACKET_TYPE_FILE_CHUNK.type:
    case PACKET_TYPE_FILE_RESEND.type:
    case PACKET_TYPE_FILE_END.type:
        file_receiver_give_packet(header->id, buf + sizeof(PacketHeader),
                                  header->length);
        break;

    default:
        break;
    }

    return 0;
}

int packet_engine_init()
{
    return 0;
}

void packet_engine_task()
{
    int len = 0;

    if (usb_available() > 0)
    {
        len = usb_receive(s_rx_buf + s_rx_buf_len, sizeof(s_rx_buf) - s_rx_buf_len);
        s_rx_buf_len += len;
        TRACE_PRINTF("Received %d bytes from USB\n", len);
    }

    // Search through the data for a packet delimiter (0x0)
    for (int i = 0; i < s_rx_buf_len; i++)
    {
        if (s_rx_buf[i] == 0x0)
        {

            // Now send the complete packet to the handler
            handle_received_packet((uint8_t *)s_rx_buf, i);

            // Shift the remaining data to the beginning of the buffer
            int remaining_start = i + 1;
            int remaining_len = s_rx_buf_len - i - 1;
            memmove(s_rx_buf, s_rx_buf + remaining_start, remaining_len);

            // Update the buffer length
            s_rx_buf_len = remaining_len;
        }
    }
}

File: software/usb/packet_engine.py
from usb import USBDev
import cobs.cobs
from packet import *
from file_receiver import file_receiver_give_packet
from gcode_sender import gcode_sender_give_packet

rx_buf = bytes()

def handle_received_packet(packet_enc):
    packet = cobs.cobs.decode(packet_enc)
    
    err = packet_check(packet)
    if err < 0:
        print(f"Received invalid packet, error: {err:d}")
        return -1
    
    header = PacketHeader.from_bytes(packet[:2])

    print(f"Received valid packet type: {header.id.packet_type:d}, len: {header.length}")
    packet_data_start = 2
    packet_data_end = 2 + header.length

    print("Packet data:")
    print(packet[packet_data_start:packet_data_end].decode('utf8', errors="replace"))

    match header.id.packet_type:
        case PACKET_TYPE_FILE_START.packet_type:
            file_receiver_give_packet(header.id, packet[packet_data_start:packet_data_end])
        case PACKET_TYPE_FILE_CHUNK.packet_type:
            file_receiver_give_packet(header.id, packet[packet_data_start:packet_data_end])
        case PACKET_TYPE_FILE_RESEND.packet_type:
            file_receiver_give_packet(header.id, packet[packet_data_start:packet_data_end])
        case PACKET_TYPE_FILE_END.packet_type:
            file_receiver_give_packet(header.id, packet[packet_data_start:packet_data_end])
        case PACKET_TYPE_GCODE.packet_type:
            gcode_sender_give_packet(header.id, packet[packet_data_start:packet_data_end])

        case _:
            pass

    return 0

def packet_engine_task(dev: USBDev):
    global rx_buf
    rx_buf_len = len(rx_buf)

    usb_data = dev.receive(1024 - rx_buf_len)

    if usb_data and len(usb_data) > 0:
        print(f"Received {len(usb_data):d} bytes from USB")
        rx_buf += usb_data
    
    for i, by in enumerate(rx_buf):
        if by == 0x00:
            # Send the complete packet
            handle_received_packet(rx_buf[:i])

            # Shift back the remaining data
            remaining_start = i + 1
            rx_buf = rx_buf[remaining_start:]
            break